Corrections in FET and Implementation of "evaluate" for tallies

2018uch0002 opened the pull request:
The PR incorporates a few corrections on previous implementations of FET. A brief description of  corrections are as follows-
1. The function `get_scaled_radius_and_angle` in `CylinderFilter` is corrected to provide the correct value of scaled radius and azimuth angle.
2. Another correction in `CylinderFilter` is made in the method to obtain the appropriate location of the cylinder in the multidimensional array of cylinder. 
3. A small bug in the `for loop` responsible for iterating over different axis in the `LengedreFET::score_collision` is corrected

The PR also includes a `virtual` method `evaluate` on  abstract class`ITally` for evaluating the tally in simulation. To achieve this, some further changes are also made on filters. The method to get the volume as inverse of volume of any mesh bin in the filter is applied. Similarly, the method to get the difference of max and min bounds in the `CartesianFilter` is applied. It will help to calculate the scaled-x, y, and z in `LegendreFET::score_collision`, `LegendreFET::scoure_source` and `Legendre::evaluate`. 

It is understood, for each particle calculating the Legendre-polynomials of 0 to higher order each time does affect the time in computation. Therefore, a `LegendrePolynomials` class is made, which will calculate and store the coefficients of each order at the time of its construction. The function `evaluate_legendres` in the class can take the `x` and simply provides the `std::vector<double>` for each order, i.e. 0 to max-give order.

HunterBelanger commented on include/tallies/general_tally.hpp:
```diff
@@ -25,6 +25,11 @@ class GeneralTally : public ITally {
 
+  virtual double evaluate(const Position r,
+                          const double E) const override final;
+  virtual std::vector<double> evaluate(const std::vector<Position> r,
+                                       const double E) const override final;
```
```suggestion
  virtual std::vector<double> evaluate(const std::vector<std::pair<Position,double>>& r_E) const override final;
```
We should have something like this, so that we could have a different energy with each position.

HunterBelanger commented on include/tallies/general_tally.hpp:
```diff
@@ -25,6 +25,11 @@ class GeneralTally : public ITally {
 
   void score_source(const BankedParticle& p) override final;
 
+  virtual double evaluate(const Position r,
+                          const double E) const override final;
```
```suggestion
  virtual double evaluate(const Position& r,
                          const double& E) const override final;
```

HunterBelanger commented on include/tallies/itally.hpp:
```diff
@@ -64,6 +64,11 @@ class ITally {
   void record_generation(double mulitplier = 1.0);
 
+  // evaluate tally
+  virtual double evaluate(const Position r, const double E) const = 0;
+  virtual std::vector<double> evaluate(const std::vector<Position> r,
+                                       const double E) const = 0;
```
```suggestion
  virtual double evaluate(const Position& r, const double& E) const = 0;
  virtual std::vector<double> evaluate(const std::vector<std::pair<Position,double>>& r_E) const = 0;
```

HunterBelanger commented on include/tallies/legendre_fet.hpp:
```diff
@@ -12,6 +13,7 @@ using StaticVector6 = boost::container::static_vector<size_t, 6>;
 
 #include <iostream>
 #include <memory>
+#include <span>
```
```suggestion
```
I don't think we are using std::span in this file, so we can remove this.

HunterBelanger replied:
Are we using std::span in here now ?

2018uch0002 replied:
I am sorry about this. we are not using it here. I don't how I missed it.

HunterBelanger commented on include/tallies/legendre_fet.hpp:
```diff
@@ -36,6 +38,10 @@ class LegendreFET : public ITally {
 
   void score_source(const BankedParticle& p) override final;
 
+  double evaluate(const Position r, const double E) const override final;
+  std::vector<double> evaluate(const std::vector<Position> positions,
+                               const double E) const override final;
```
These should be changed to the new signature.

HunterBelanger commented on include/tallies/position_filter.hpp:
```diff
@@ -32,6 +32,10 @@ class PositionFilter {
   virtual StaticVector3 get_shape() const = 0;
 
+  // method to get the volume and inverse of the volume
+  virtual double dV(StaticVector3 index) const = 0;
+  virtual double inv_dV(StaticVector3 index) const = 0;
```
```suggestion
  virtual double dV(const StaticVector3& index) const = 0;
  virtual double inv_dV(const StaticVector3& index) const = 0;
```

2018uch0002 replied:
As discussed last time, I have removed the functions to get the volumes and inverse of it. Therefore, we will not be calling these functions anywhere else.

HunterBelanger commented on include/tallies/regular_cartesian_mesh_filter.hpp:
```diff
@@ -16,12 +16,30 @@ class RegularCartesianMeshFilter : public CartesianFilter {
+  double dz(const StaticVector3& /*index*/) const override final { return dz_; }
+  double inv_dz(const StaticVector3& /*index*/) const override final {
+    return dz_inv_;
+  }
+  double dV(StaticVector3 /*index*/) const override final {
```
```suggestion
  }
  double dV(const StaticVector3& /*index*/) const override final {
```

HunterBelanger commented on include/tallies/regular_cartesian_mesh_filter.hpp:
```diff
@@ -16,12 +16,30 @@ class RegularCartesianMeshFilter : public CartesianFilter {
+  double dV(StaticVector3 /*index*/) const override final {
+    return dx_ * dy_ * dz_;
+  }
+  double inv_dV(StaticVector3 /*index*/) const override final {
```
```suggestion
  double inv_dV(const StaticVector3& /*index*/) const override final {
```

HunterBelanger commented on include/tallies/legendre_polynomial.hpp:
```diff
@@ -0,0 +1,31 @@
+  std::vector<double> evaluate_legendres(const double x) const;
+
+ private:
+  std::vector<std::vector<double>> legendre_coeff_;
```
We should use an `xt::xtensor<double, 2>` for this instead of a vector of vectors. It will be much faster and nicer to our CPU cache. We should also make this change in the ZernikePolynomial if it is also there.

We also might want to consider templating on the order and then making legendre_coeff_ a static variable, that way it is only stored once, but let's make the other changes first.

2018uch0002 replied:
Instead of 2D dimension, I have incorporated in one-dimension and altered the class accordingly. I can confirm that the evaluation of Legendre and Zernike Polynomial is working correctly.

HunterBelanger commented on include/tallies/zernike_fet.hpp:
```diff
@@ -36,6 +39,13 @@ class ZernikeFET : public ITally {
+  std::vector<double> evaluate_FET(std::span<Position> coordinates,
+                                   const double En) const;
+
+  double evaluate(const Position r, const double E) const override final;
+  std::vector<double> evaluate(const std::vector<Position> positions,
+                               const double E) const override final;
```
Need to update these signatures.

2018uch0002 replied:
I have updated at all places.

HunterBelanger commented on include/tallies/zernike_fet.hpp:
```diff
@@ -4,18 +4,21 @@
 
 #include <yaml-cpp/yaml.h>
 
+#include <span>
+
```
```suggestion
```

HunterBelanger replied:
Are we using std::spand here now ?

2018uch0002 replied:
I am sorry about this. we are not using it here. I don't how I missed it.

HunterBelanger commented on include/tallies/zernike_fet.hpp:
```diff
@@ -36,6 +39,13 @@ class ZernikeFET : public ITally {
 
   void score_source(const BankedParticle& p) override final;
 
+  std::vector<double> evaluate_FET(std::span<Position> coordinates,
+                                   const double En) const;
```
This should be removed, right ?

2018uch0002 replied:
Yes, I have removed it.

HunterBelanger commented on include/tallies/zernike_polynomial.hpp:
```diff
@@ -23,6 +23,9 @@ class ZernikePolynomials {
+
  private:
   std::size_t order_, max_n_;
   std::vector<std::vector<double>> Zr_coefficients_;
```
We should probably move this to be an `xt::xtensor<double, 2>`.

HunterBelanger commented on include/tallies/zernike_polynomial.hpp:
```diff
@@ -23,6 +23,9 @@ class ZernikePolynomials {
                                         const double theta) const;
 
+  double orthonormalization_constant(const std::size_t& order,
+                                     bool is_pi_multiply = true) const;
```
Maybe rename the `is_pi_multiply` variable. It is not clear to me whether this means we need it multiplied by pi or not. Also, is this option even necessary ? We should try to keep the data as consistent as possible and not allow such a variation if possible.

2018uch0002 replied:
I have realised that it can create confusion. So, I have removed it.

HunterBelanger commented on src/general_tally.cpp:
```diff
@@ -206,6 +206,98 @@ void GeneralTally::score_source(const BankedParticle& p) {
+      inv_dV = position_filter_->inv_dV(position_indices);
+    }
+  }
+  const double tally_value =
+      inv_dV * tally_avg_.element(indices.begin(), indices.end());
```
So we should not be normalizing by the volume here. The reason for this is that it will not always be possible with future position filters. For example, if we have a position filter based on the Cell or Material, we have no way of knowing their volume, and this won't be possible. We should therefore be returning the volume integrated tallies, i.e. V<flx>, and if the user wants <flx> in the region, it is up to them to find V and normalize.

HunterBelanger commented on src/general_tally.cpp:
```diff
@@ -206,6 +206,98 @@ void GeneralTally::score_source(const BankedParticle& p) {
+      }
+    }
+
+    const double tally_value =
+        inv_dV * tally_avg_.element(indices.begin(), indices.end());
```
Again, no volume normalization.

HunterBelanger commented on src/general_tally.cpp:
```diff
@@ -206,6 +206,98 @@ void GeneralTally::score_source(const BankedParticle& p) {
 }
 
+// method to evaluate the tally
+double GeneralTally::evaluate(const Position r, const double E) const {
```
I am realizing that once we add an `energy_out_` filter in this class, we will also need Eout to be able to evaluate the tally... We need to think about how we want to handle this...

HunterBelanger commented on src/legendre_fet.cpp:
```diff
@@ -7,6 +7,8 @@
 #include <boost/container/static_vector.hpp>
 using StaticVector6 = boost::container::static_vector<size_t, 6>;
 
+#include <span>
+
```
```suggestion
```

HunterBelanger commented on src/legendre_fet.cpp:
```diff
@@ -196,40 +197,259 @@ void LegendreFET::score_source(const BankedParticle& p) {
+  }
+
+  // divide the value by the volume of the bin
+  tally_value *= inv_dx_ * inv_dy_ * inv_dz_;
```
Again, we probably don't want to normalize by the volume.

HunterBelanger commented on src/legendre_fet.cpp:
```diff
@@ -196,40 +197,259 @@ void LegendreFET::score_source(const BankedParticle& p) {
 
-    it_axis++;
+    // divide the value by the volume of the bin
+    tally_value *= inv_dx_ * inv_dy_ * inv_dz_;
```
No volume normalization.

HunterBelanger commented on src/legendre_polynomial.cpp:
```diff
@@ -0,0 +1,69 @@
+
+std::vector<double> LegendrePolynomials::evaluate_legendres(
+    const double x) const {
+  // get the powers upto that orders
```
```suggestion
  // get the powers up to that order
```

HunterBelanger commented on src/legendre_polynomial.cpp:
```diff
@@ -0,0 +1,69 @@
+    x_powers.push_back(x_pow);
+    x_pow *= x;
+  }
+  // evalute the legendre polynomails of each order
```
```suggestion
  // evaluate the legendre polynomials of each order
```

HunterBelanger commented on src/legendre_polynomial.cpp:
```diff
@@ -0,0 +1,69 @@
+std::vector<double> LegendrePolynomials::evaluate_legendres(
+    const double x) const {
+  // get the powers upto that orders
+  std::vector<double> x_powers;
+  x_powers.reserve(order_ + 1);
+  double x_pow = 1.;
+  for (std::size_t i = 0; i <= order_; i++) {
+    x_powers.push_back(x_pow);
+    x_pow *= x;
+  }
```
Do we really need to create a vector to store these ? Would it be possible to directly compute the needed x_pow in the loop bellow ? This would likely go a long way to speeding up the evaluation.

HunterBelanger commented on src/zernike_polynomial.cpp:
```diff
@@ -118,4 +117,17 @@ std::vector<double> ZernikePolynomials::evaluate_zernikes(
+  std::pair<int, int> n_and_l = get_n_and_l(order);
+  const double n = static_cast<double>(n_and_l.first);
+  const double l = static_cast<double>(n_and_l.second);
+  if (l == 0) {
+    return is_pi_multiply ? PI / (n + 1.) : 1. / (n + 1.);
+  }
+  return is_pi_multiply ? 0.5 * PI / (n + 1.) : 0.5 / (n + 1.);
```
We do we only need the pi some times ?

HunterBelanger commented on src/zernike_fet.cpp:
```diff
@@ -295,6 +311,191 @@ void ZernikeFET::score_source(const BankedParticle& p) {
+  // get the inverse of radius for area
+  const double inv_dV = cylinder_filter_->dV(cylinder_index);
+  // divide the tally_value by volume
+  tally_value *= tally_zr * inv_dV;
```
Again, no volume normalization.

HunterBelanger commented on src/zernike_fet.cpp:
```diff
@@ -295,6 +311,191 @@ void ZernikeFET::score_source(const BankedParticle& p) {
+    // get the inverse of radius for area
+    const double inv_dV = cylinder_filter_->dV(cylinder_index);
+    // divide the tally_value by volume
+    tally_value *= tally_zr * inv_dV;
```
Again, no volume normalization.

HunterBelanger reviewed (commented):
Here are some initial comments on the work. We definitely need to think about how we want to handle the tally evaluation on `GeneralTally` in the future. Once we include things like an outgoing energy filter or a legendre moment filter, it will be complicated fast. This just might be a place where we will have to treat the `GeneralTally` differently from the FET tallies.

HunterBelanger commented on include/tallies/legendre_polynomial.hpp:
```diff
@@ -0,0 +1,41 @@
+  LegendrePolynomials() = default;
+  LegendrePolynomials(std::size_t order);
+
+  //~LegendrePolynomials() = default;
```
```suggestion
```

HunterBelanger commented on include/tallies/legendre_polynomial.hpp:
```diff
@@ -0,0 +1,41 @@
+//
+class LegendrePolynomials {
+ public:
+  LegendrePolynomials() = default;
```
Do we need a default constructor for this class ? I don't like the idea of this being uninitialized like this. By default, it would have an order of 0, which is of no use. We should probably delete this default constructor.

2018uch0002 replied:
You are right. I have changed it.

HunterBelanger commented on include/tallies/legendre_polynomial.hpp:
```diff
@@ -0,0 +1,41 @@
+
+  // function to evaluate the factorial calculations for coefficients for given
+  // order and term
+  double coeff_factorial_evaluation(const std::size_t& n,
```
This can be const, right ?

2018uch0002 replied:
Yes, I have changed this.

HunterBelanger commented on include/tallies/zernike_polynomial.hpp:
```diff
@@ -23,19 +23,21 @@ class ZernikePolynomials {
 
   // function for the factorial
-  std::size_t factorial(std::size_t N) {
+  double factorial(std::size_t N) {
```
```suggestion
  double factorial(std::size_t N) const {
```

2018uch0002 replied:
Changes are made here.

HunterBelanger commented on include/tallies/zernike_polynomial.hpp:
```diff
@@ -48,7 +50,7 @@ class ZernikePolynomials {
       l = m * sign_change;
       // get n
-      n = static_cast<int>(-1. + std::sqrt(2 * order + 1 - l));
+      n = static_cast<int>(-1. + std::sqrt(2. * static_cast<double>(order) + 1. - static_cast<double>(l)));
```
I find it scarry that we are casting the result of std::sqrt to an int. Do we need to perform this cast, or can we keep the double here ?

2018uch0002 replied:
Then, can I use the std::floor ?

HunterBelanger commented on include/tallies/zernike_polynomial.hpp:
```diff
@@ -48,7 +50,7 @@ class ZernikePolynomials {
-      n = static_cast<int>(-1. + std::sqrt(2 * order + 1 - l));
+      n = static_cast<int>(-1. + std::sqrt(2. * static_cast<double>(order) + 1. - static_cast<double>(l)));
       // check if selected and m and n are correct or not
       if (j == 0.5 * (n * (n + 2) + l)) {
```
This is not a safe comparison, as we are comparing an int to a double (due to the 0.5), and this is leading to an implicit cast. This should be re-written.

HunterBelanger commented on src/general_tally.cpp:
```diff
@@ -206,6 +206,92 @@ void GeneralTally::score_source(const BankedParticle& p) {
+    }
+  }
+  const double tally_value = tally_avg_.element(indices.begin(), indices.end());
+  return tally_value;
```
```suggestion

  return tally_avg_.element(indices.begin(), indices.end());
```

HunterBelanger commented on src/general_tally.cpp:
```diff
@@ -206,6 +206,92 @@ void GeneralTally::score_source(const BankedParticle& p) {
+        tally_avg_.element(indices.begin(), indices.end());
+    tallied_values.push_back(tally_value);
+  }
+  return tallied_values;
```
```suggestion
    const double tally_value =
        tally_avg_.element(indices.begin(), indices.end());
    tallied_values.push_back(tally_value);
  }
  
  return tallied_values;
```

HunterBelanger commented on src/legendre_fet.cpp:
```diff
@@ -196,40 +195,255 @@ void LegendreFET::score_source(const BankedParticle& p) {
+          }
+          tally_value *= fet_value_z;
+        }
+        default:
+          break;
```
Because we are using an enum class here, I don't think we need the default. If we do eventually need the default though, I think we should throw a fatal_error here, as we should never get here.

2018uch0002 replied:
Should I throw the `fatal_error` in other functions like `score_collision` as well?

2018uch0002 replied:
I have removed the default from all the other places as well.

HunterBelanger commented on src/legendre_polynomial.cpp:
```diff
@@ -0,0 +1,83 @@
+      coeff_length =
+          length * (length + 1) + static_cast<std::size_t>(static_cast<double>(order_) * 0.5 + 1);
+    } else {
+      const std::size_t length = static_cast<std::size_t>(static_cast<double>(order_) * 0.5 + 1);
```
Here, order_ isn't even, and we are still multiplying by 0.5. Are we sure this cast from double to std::size_t is safe ?

HunterBelanger commented on src/legendre_polynomial.cpp:
```diff
@@ -0,0 +1,83 @@
+  }
+  legendre_coeff_.reserve(coeff_length);
+
+  double inv_2 = 1.;
```
Is this supposed to be 1. / 2. ? Also, I think this can be made const.

2018uch0002 replied:
I think, I should rename the variable from `double inv_2` to `double inv_2n` as this tells the inverse of the 2 to the power polynomial-order. The variable will be modified by multiplying 0.5 into for next order calculations. So, we cannot make it `const`.

HunterBelanger commented on src/legendre_polynomial.cpp:
```diff
@@ -0,0 +1,83 @@
+
+  std::vector<double> sol = evaluate_legendres(0.1);
+  for (std::size_t i = 0; i < sol.size(); i++) {
+    std::cout << "order = " << i << "\t\t" << sol[i] << "\n";
```
We should not have an std::cout here.

2018uch0002 replied:
I apologies here. It should not be here.

HunterBelanger commented on src/zernike_fet.cpp:
```diff
@@ -295,6 +310,185 @@ void ZernikeFET::score_source(const BankedParticle& p) {
+  }
+  // create a temperaroy tracker for getting a position index
+  Direction dir;
+  Tracker trkr(r, dir);
+  // get the positional indexes from cylinder_filter
+  StaticVector3 cylinder_index = cylinder_filter_->get_indices(trkr);
```
Maybe for cylinder filters and cartesian filters, we can create a special method that can return the indices with just the position ? That would allow us to avoid the creation of this temporary Tracker, which should greatly reduce the computational costs. Because creating a tracker in this case is a lot of computational work and it is not purely necessary for these two types of position filters. We couldn't use this on a general tally, as those could have any type of position filter, but here we know what we are getting. This could also be done for the LegendreFET.

HunterBelanger commented on src/zernike_fet.cpp:
```diff
@@ -295,6 +310,185 @@ void ZernikeFET::score_source(const BankedParticle& p) {
+  // get the inverse of cylinder length for scaled-z and volume calculation
+  const double inv_dz_ = cylinder_filter_->inv_dz();
+  // second- evaluate the legendre if exist
+  if (check_for_legendre == true) {
```
```suggestion
  if (check_for_legendre) {
```

HunterBelanger commented on src/zernike_fet.cpp:
```diff
@@ -295,6 +310,185 @@ void ZernikeFET::score_source(const BankedParticle& p) {
+    // get the inverse of cylinder length for scaled-z and volume calculation
+    const double inv_dz_ = cylinder_filter_->inv_dz();
+    // second- evaluate the legendre if exist
+    if (check_for_legendre == true) {
```
```suggestion
    if (check_for_legendre) {
```

HunterBelanger commented on src/zernike_polynomial.cpp:
```diff
@@ -1,12 +1,30 @@
 ZernikePolynomials::ZernikePolynomials(std::size_t order)
     : order_(order), max_n_(), Zr_coefficients_(), m_(), n_(), Zr_types_() {
   // Evaluate the polynomial coefficients up to given order
   // Zr_coefficients_ will be having the coeff. stating from lowest power to
```
```suggestion
  // Zr_coefficients_ will have the coeffs. stating from lowest power to
```

HunterBelanger commented on src/zernike_polynomial.cpp:
```diff
@@ -1,12 +1,30 @@
-  Zr_coefficients_.reserve(order_ + 1);
+
+  // get a maximum size to store the coefficients
+  std::pair<int, int> high_n_and_l = get_n_and_l(order_);
+  // store the n for highest order
+  // this is required as we have to evaluate upto this power
+  max_n_ = static_cast<std::size_t>(high_n_and_l.first);
```
Why do we immediately cast these ints to an std::size_t ? Can the method get_n_and_l immediately return a pair of std::size_t ?

HunterBelanger commented on src/zernike_polynomial.cpp:
```diff
@@ -31,27 +49,23 @@ ZernikePolynomials::ZernikePolynomials(std::size_t order)
     for (std::size_t i = 0; i <= max_k; i++) {
       k = max_k - i;
-      const double numeriator = static_cast<double>(factorial(n - k));
+      const double numeriator = factorial(static_cast<std::size_t>(n - static_cast<int>(k)));
```
Why the cast to in here ? I thought factorial took an std::size_t ?

HunterBelanger commented on src/zernike_polynomial.cpp:
```diff
@@ -31,27 +49,23 @@ ZernikePolynomials::ZernikePolynomials(std::size_t order)
     std::vector<double> single_order_Zr_coeff;
     single_order_Zr_coeff.reserve(max_k + 1);
     double sign_change = 1.;
-
     if (static_cast<int>(max_k % 2) == 1) sign_change = -1.;
 
     std::size_t k;
     for (std::size_t i = 0; i <= max_k; i++) {
       k = max_k - i;
-      const double numeriator = static_cast<double>(factorial(n - k));
+      const double numeriator = factorial(static_cast<std::size_t>(n - static_cast<int>(k)));
       const std::size_t avg_n_m = static_cast<std::size_t>((n + m) * 0.5);
-      const std::size_t mid_n_m = static_cast<std::size_t>((n - m) * 0.5);
-      const double denominator = static_cast<double>(
-          factorial(k) * factorial(avg_n_m - k) * factorial(mid_n_m - k));
+      const std::size_t mid_n_m = static_cast<std::size_t>(static_cast<double>(n - m) * 0.5);
+      const double denominator =
```
We are sure these two casts are safe here ?

HunterBelanger commented on src/zernike_polynomial.cpp:
```diff
@@ -61,16 +75,30 @@ double ZernikePolynomials::evaluate_zernike_at_order(
+    for (std::size_t ord = 0; ord < order; ord++) {
+      std::pair<int, int> n_and_l = get_n_and_l(ord);
+      const std::size_t n = static_cast<std::size_t>(n_and_l.first);
+      const std::size_t m = static_cast<std::size_t>(std::abs(n_and_l.second));
```
Why are we changing the variable name from l to m here ? This is very confusing.

HunterBelanger commented on src/zernike_polynomial.cpp:
```diff
@@ -61,16 +75,30 @@ double ZernikePolynomials::evaluate_zernike_at_order(
+    }
+    std::pair<int, int> n_and_l = get_n_and_l(order);
+    const std::size_t n = static_cast<std::size_t>(n_and_l.first);
+    const std::size_t m = static_cast<std::size_t>(std::abs(n_and_l.second));
```
Again, why are we calling l m ?

HunterBelanger commented on src/zernike_polynomial.cpp:
```diff
@@ -61,16 +75,30 @@ double ZernikePolynomials::evaluate_zernike_at_order(
+    value += Zr_coefficients_[i] * x_k;
     x_k *= x_k2;
   }
   double mtheta = static_cast<double>(m) * theta;
```
const ?

2018uch0002 replied:
I am really not sure, what exactly you refereeing to.

HunterBelanger commented on src/zernike_polynomial.cpp:
```diff
@@ -96,17 +124,21 @@ std::vector<double> ZernikePolynomials::evaluate_zernikes(
+      k += 2;
+      it_coeff++;
     }
     double mtheta = static_cast<double>(m) * theta;
```
const ?

HunterBelanger reviewed (commented):
Here are a new set of comments. I definitely think we should add a new get indices method to the cylindrical and cartesian filters that only takes a position. This will greatly facilitate our use case here, and lead to better performance.

2018uch0002 commented:
I have made the changes, and important ones are described below:

1. If you remember, the `LegendrePolynomials` class was made to evaluate the polynomials up to given order. This class is now removed. Instead of this, the recursive relation is used at the time of looping over all order to compute and store the coefficients of FET. This clearly removed the overhead of storing the values in the `std::vector<double>` while calling the evaluate-functions, which we had earlier. The time-cost is very similar for both implementations. The same methodology is at all the places in the tallies where Legendre Polynomials are required.

2.  The `LegendreFET::evaluate` function has been rearranged, which is very similar to `LegendreFET::score_collision`. I believe doing this, all the functions are consistent with each other and easy to understand.

3. You have mentioned earlier that the we should not create the `Tracker` in `LegendreFET::evaluate` and `ZernikeFET::evaluate`  only to get the position-index and you asked me to create the same method in the `CartesianFilter` and `CylinderFilter`. I have incorporated this by making the function `get_position_index` for the mentioned filters.

HunterBelanger commented on include/tallies/zernike_polynomial.hpp:
```diff
@@ -17,56 +17,36 @@ class ZernikePolynomials {
-  std::size_t factorial(std::size_t N) {
+  double factorial(std::size_t N) const {
     if (N == 1 || N == 0)
       return 1;
```
```suggestion
      return 1.;
```

HunterBelanger commented on src/regular_cartesian_mesh_filter.cpp:
```diff
@@ -76,6 +76,22 @@ StaticVector3 RegularCartesianMeshFilter::get_indices(
+  StaticVector3 indices;
+  int index_x = static_cast<int>(std::floor((r.x() - r_low_.x()) * dx_inv_));
+  int index_y = static_cast<int>(std::floor((r.y() - r_low_.y()) * dy_inv_));
+  int index_z = static_cast<int>(std::floor((r.z() - r_low_.z()) * dz_inv_));
```
```suggestion
  const int index_x = static_cast<int>(std::floor((r.x() - r_low_.x()) * dx_inv_));
  const int index_y = static_cast<int>(std::floor((r.y() - r_low_.y()) * dy_inv_));
  const int index_z = static_cast<int>(std::floor((r.z() - r_low_.z()) * dz_inv_));
```
